From 00244b40ad1b10f19f57acd7ff22dd8d27964a97 Mon Sep 17 00:00:00 2001
From: Marcel Apfelbaum <marcel.a@redhat.com>
Date: Thu, 20 Nov 2014 11:25:31 +0100
Subject: [PATCH 1/2] hw/pci: fixed error flow in pci_qdev_init

Message-id: <1416482732-28870-2-git-send-email-marcel.a@redhat.com>
Patchwork-id: 62489
O-Subject: [RHEL-7.1 qemu-kvm-rhev PATCH 1/2] hw/pci: fixed error flow in pci_qdev_init
Bugzilla: 1166067
RH-Acked-by: Eric Blake <eblake@redhat.com>
RH-Acked-by: Amos Kong <akong@redhat.com>
RH-Acked-by: Markus Armbruster <armbru@redhat.com>

Verify return code for pci_add_option_rom.

Signed-off-by: Marcel Apfelbaum <marcel.a@redhat.com>
Reviewed-by: Michael S. Tsirkin <mst@redhat.com>
Signed-off-by: Michael S. Tsirkin <mst@redhat.com>
Reviewed-by: Markus Armbruster <armbru@redhat.com>
(cherry picked from commit 178e785fb4507ec3462dc772bbe08303416ece47)
Signed-off-by: Marcel Apfelbaum <marcel.a@redhat.com>
Signed-off-by: Miroslav Rezanina <mrezanin@redhat.com>
---
 hw/pci/pci.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/hw/pci/pci.c b/hw/pci/pci.c
index 42995e6..a227e95 100644
--- a/hw/pci/pci.c
+++ b/hw/pci/pci.c
@@ -1777,7 +1777,12 @@ static int pci_qdev_init(DeviceState *qdev)
         pci_dev->romfile = g_strdup(pc->romfile);
         is_default_rom = true;
     }
-    pci_add_option_rom(pci_dev, is_default_rom);
+
+    rc = pci_add_option_rom(pci_dev, is_default_rom);
+    if (rc != 0) {
+        pci_unregister_device(DEVICE(pci_dev));
+        return rc;
+    }
 
     return 0;
 }
-- 
1.8.3.1

